src/cache/cache.go

```diff
@@ -13,32 +13,35 @@ import (
 var log = logging.MustGetLogger("cache")
 
 // NewCache is the factory function for creating a cache setup from the given config.
-func NewCache(config *core.Configuration) core.Cache {
-	c := newSyncCache(config, false)
-	if config.Cache.Workers > 0 {
-		return newAsyncCache(c, config)
+func NewCache(state *core.BuildState) core.Cache {
+	c := newSyncCache(state, false)
+	if state.Config.Cache.Workers > 0 {
+		return newAsyncCache(c, state.Config)
 	}
 	return c
 }
 
 // newSyncCache creates a new cache, possibly multiplexing many underneath.
-func newSyncCache(config *core.Configuration, remoteOnly bool) core.Cache {
+func newSyncCache(state *core.BuildState, remoteOnly bool) core.Cache {
 	mplex := &cacheMultiplexer{}
-	if config.Cache.Dir != "" && !remoteOnly {
-		mplex.caches = append(mplex.caches, newDirCache(config))
+	if state.Config.Cache.Dir != "" && !remoteOnly {
+		mplex.caches = append(mplex.caches, newDirCache(state.Config))
 	}
-	if config.Cache.RPCURL != "" {
-		cache, err := newRPCCache(config)
+	if state.Config.Remote.URL != "" {
+		mplex.caches = append(mplex.caches, newRemoteCache(state))
```

Comment:
Feels a little odd this takes the whole state when the other caches take the config - what does the remote cache setup need from the state at this point?

Reply:
It needs the PathHasher since it needs to hash files individually (just having the overall target hash isn't granular enough for the API). That lives on the state so it doesn't get persisted between subsequent runs within the process (e.g. for `plz watch` or `plz query changed` etc).